### PR TITLE
add `SelectableEventedList`

### PR DESCRIPTION
**Pull request by alisterburt**
This PR adds the `SelectableEventedList` from napari - a `MutableSequence` with a built in selection model

It's an almost straight [port from napari](https://github.com/napari/napari/blob/main/napari/utils/events/containers/_selectable_list.py) with the addition of...
- trying to keep mypy happy 
- elements removed from a list in `SEL.remove_selected()` are also removed from the selection and returned from the method
- added a `wraparound` option to `SEL.select_next()` and `SEL.select_previous()`

will leave a few comments below on some issues I had with mypy/events. As always no rush to get to this from my side 🙂

### CODE REVIEW

File: tests/containers/test_selectable_evented_list.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    test_list.selection.discard = Mock(wraps=test_list.selection.discard)
+    test_list.selection = {0}
+    assert 0 in test_list.selection
+    test_list.remove(0)
+    assert 0 not in test_list.selection
+    test_list.selection.discard.assert_called_once()
+
+
+def test_remove_selected(test_list):
+    """Test that remove selected method removes all selected elements"""
```
Comment (alisterburt) [inline tests/containers/test_selectable_evented_list.py:108]:
```suggestion
    """Test items are removed from both the selection and the list."""
```

File: psygnal/containers/_selectable_evented_list.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class SelectableEventedList(Selectable[_T], EventedList[_T]):
+    """List object with a built in selection model."""
+
+    def __init__(self, *args: _T, **kwargs: _T) -> None:
+        self._activate_on_insert: bool = True
+        super().__init__(*args, **kwargs)
+        self.events.removed.connect(self.selection.discard)
+        self.selection._pre_add_hook = self._pre_select_hook  # type:ignore
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply psygnal/containers/_selectable_evented_list.py:17]: yes, proper way to ignore... but I never liked how i did that.  see if you can come up with a way that doesn't require assignment to a method?

File: psygnal/containers/_selectable_evented_list.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    def select_previous(
+        self, expand_selection: bool = False, wraparound: bool = False
+    ) -> None:
+        """Select the previous item in the list."""
+        self.select_next(
+            step=-1, expand_selection=expand_selection, wraparound=wraparound
+        )
+
+    def remove_selected(self) -> Tuple[Any, ...]:
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply psygnal/containers/_selectable_evented_list.py:69]: did you try `Tuple[_T, ...]` as well? (with the ellipis?)

File: psygnal/containers/_selectable_evented_list.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alisterburt) [reply psygnal/containers/_selectable_evented_list.py:17]: tried something but don't think it's cleaner (not quite working yet)

```python
class SelectableEventedList(Selectable[_T], EventedList[_T]):
    """List object with a built in selection model."""

    def __init__(self, *args: _T, **kwargs: _T) -> None:
        self._activate_on_insert: bool = True
        super().__init__(*args, **kwargs)
        self.events.removed.connect(self.selection.discard)
        self._establish_pre_select_hook()

    def _pre_select_hook(self, value: _T) -> _T:
        """Check an item is in the list before adding to the selection."""
        if value not in self:
            raise ValueError("Cannot select an item that is not in the list.")
        return value

    def _establish_pre_select_hook(self):
        """Override the pre-add hook on the selection with another method."""
        def override(func, new_function):
            def inner(*args, **kwargs):
                new_function(*args, **kwargs)
            return inner
        self.selection._pre_add_hook = override(
            self.selection._pre_add_hook, self._pre_select_hook
        )
```

Have a few other ideas in mind but they seem a bit convoluted too
- a `hookmethod` decorator which provides an override method on the function itself

**Comment by alisterburt** [comment]
friendly ping 🙂 

I've come back to this a few times and this is the best way I could come up with to remove the method assignment -> it adds the restriction that `Selectable` can now only be used with `Container` types 

I'm not sure how to fix the typing errors in `test_containers.yml` so any advice there would be appreciated! pre-commit fail seems unrelated

**Comment by tlambert03** [comment]
I like it!!  the _parent solution is much nicer than before.  lets go with it.  Will finish up reviewing other little issues,  might post some documentation changes, and then will merge later

File: psygnal/containers/_selectable_evented_list.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        """Remove selected items from the list and the selection."""
+        selected_items = tuple(self.selection)
+        idx = 0
+        for item in list(self.selection):
+            idx = self.index(item)
+            self.remove(item)
+            # shouldn't be necessary but remove is not discarding from selection
+            # when called from here...
+            self.selection.discard(item)
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply psygnal/containers/_selectable_evented_list.py]: figured it out... the issue was the connection above that did: `self.events.removed.connect(self.selection.discard)`

`events.removed` emits _2_ arguments, `(index, object)`, whereas discard only wants the object.  It happily accepted the first argument instead (just the index), and so was removing the wrong stuff.  changed it to `        self.events.removed.connect(lambda _, b: self.selection.discard(b))` and now it works

File: tests/containers/test_selectable_evented_list.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        ({0}, 2, True, False, {0, 2}),
+        ({0, 1}, 1, False, False, {1}),
+        ({0}, 5, False, False, {4}),
+        ({0}, 5, False, True, {0}),
+    ],
+)
+def test_select_next(
+    test_list, initial_selection, step, expand_selection, wraparound, expected
+):
+    """Test select next method behaviour."""
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply tests/containers/test_selectable_evented_list.py:75]: it's fine :)  test speaks for itself

File: psygnal/containers/_selectable_evented_list.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alisterburt) [reply psygnal/containers/_selectable_evented_list.py]: gah! thanks for digging - silly mistake, glad you found it!

**Comment by alisterburt** [comment]
this is a vibe 😆 
image

**Review by tlambert03 — APPROVED** [review]
Thanks for this @alisterburt!  Tests look fantastic.  Will merge as soon as everything passes

**Comment by alisterburt** [comment]
🥳 excellent stuff, pleasure doing business, thanks!!

**Pull request merged**